logsearchapi/server/server.go

```diff
@@ -0,0 +1,152 @@
+// +build go1.13
+
+/*
+ * Copyright (C) 2020, MinIO, Inc.
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+package server
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/http/httputil"
+)
+
+// LogSearch represents the Log Search API server
+type LogSearch struct {
+	// Configuration
+	PGConnStr                      string
+	AuditAuthToken, QueryAuthToken string
+	MaxRetentionMonths             int
+
+	// Runtime
+	DBClient *DBClient
+	*http.ServeMux
+}
+
+// NewLogSearch creates a LogSearch
+func NewLogSearch(pgConnStr, auditAuthToken string, queryAuthToken string, maxRetentionMonths int) (ls *LogSearch, err error) {
+	ls = &LogSearch{
+		PGConnStr:          pgConnStr,
+		AuditAuthToken:     auditAuthToken,
+		QueryAuthToken:     queryAuthToken,
+		MaxRetentionMonths: maxRetentionMonths,
+	}
+
+	// Initialize DB Client
+	ls.DBClient, err = NewDBClient(context.Background(), ls.PGConnStr)
+	if err != nil {
+		// FIXME(aditya): Remove connection string as it contains creds
+		return nil, fmt.Errorf("Error connecting to db: %v, connection string: %s", err, ls.PGConnStr)
+	}
+
+	// Initialize tables in db
+	err = ls.DBClient.InitDBTables(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("Error initializing tables: %v", err)
+	}
+
+	// Initialize muxer
+	ls.ServeMux = http.NewServeMux()
+	ls.HandleFunc("/status", func(w http.ResponseWriter, r *http.Request) {})
+	ls.HandleFunc("/api/ingest", authorize(ls.ingestHandler, ls.AuditAuthToken))
+	ls.HandleFunc("/api/query", authorize(ls.queryHandler, ls.QueryAuthToken))
```

Comment:
Should we version the API endpoint? e.g. `/api/v1/query`

Reply:
Yes, we will do this subsequently. Just want to get this in, to make further changes small and easy to review.